### PR TITLE
[Serve] [2/3 Wrappers] Add Basic HTTP Adapters

### PR DESCRIPTION
**Pull request by simon-mo**
## Why are these changes needed?
This PR adds two FastAPI dependency injection function that will convert HTTP input to DataBatchType.

Depends on 
- https://github.com/ray-project/ray/pull/22913

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/serve/http_adapters.py [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+
+
+class NdArray(BaseModel):
+    """Schema for numeric array input."""
+
+    array: Union[_1DArray, _2DArray, _3DArray, _4DArray] = Field(
+        ...,
+        description=(
+            "The array content as a list or nested list. "
+            "You can pass in 1D to 4D array in its original form or the flatten verion."
```
Comment (edoakes) [inline python/ray/serve/http_adapters.py:24]:
```suggestion
            "You can pass in a 1D to 4D array in its original form or the flattened verion."
```

Unclear what original form vs. flattened version means to me here. Give an example?

File: python/ray/serve/http_adapters.py [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+        ),
+    )
+    shape: Optional[List[int]] = Field(
+        default=None,
+        description=("The shape of the array. If present, the array will be reshaped."),
+    )
+    dtype: Optional[str] = Field(
```
Comment (edoakes) [inline python/ray/serve/http_adapters.py:31]: what are the available values?

Comment (edoakes) [reply python/ray/serve/http_adapters.py:31]: can we make it an enum?

File: python/ray/serve/http_adapters.py [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+
+
+_1DArray = List[float]
+_2DArray = List[List[float]]
+_3DArray = List[List[List[float]]]
+_4DArray = List[List[List[List[float]]]]
+
+
+class NdArray(BaseModel):
+    """Schema for numeric array input."""
```
Comment (edoakes) [inline python/ray/serve/http_adapters.py:18]: is this numpy-specific btw? if so, can we call it that for clarity? `NumpyNdarray`

File: python/ray/serve/http_adapters.py [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+        description=(
+            "The dtype of the array. If present, the array will be cast by `astype`."
+        ),
+    )
+
+
+@require_packages(["numpy"])
+def array_to_databatch(payload: NdArray) -> DataBatchType:
+    """Accept a NdArray from HTTP body, returns array."""
+    import numpy as np
```
Comment (edoakes) [inline python/ray/serve/http_adapters.py:42]: why delayed import?

Comment (edoakes) [reply python/ray/serve/http_adapters.py:42]: isn't numpy a dep of ray?

File: python/ray/serve/http_adapters.py [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+        default=None,
+        description=(
+            "The dtype of the array. If present, the array will be cast by `astype`."
+        ),
+    )
+
+
+@require_packages(["numpy"])
+def array_to_databatch(payload: NdArray) -> DataBatchType:
+    """Accept a NdArray from HTTP body, returns array."""
```
Comment (edoakes) [inline python/ray/serve/http_adapters.py:41]:
```suggestion
    """Accepts an NdArray from an HTTP body and converts it to a DataBatchType."""
```

File: python/ray/serve/utils.py [hunk 6]
```diff
@@ -344,3 +346,40 @@ def __getattr__(self, key: str):
...
+            if not hasattr(func, "_require_packages_checked"):
+                missing_packages = []
+                for package in packages:
+                    try:
+                        importlib.import_module(package)
+                    except ModuleNotFoundError:
+                        missing_packages.append(package)
+                if len(missing_packages) > 0:
+                    raise ImportError(
+                        f"{func} requires {packages} to run but "
```
Comment (edoakes) [inline python/ray/serve/utils.py:375]:
```suggestion
                        f"{func} requires packages {packages} to run but "
```

File: python/ray/serve/utils.py [hunk 7]
```diff
@@ -344,3 +346,40 @@ def __getattr__(self, key: str):
...
+                        f"{missing_packages} are missing. Please "
+                        "`pip install` them or add them to "
+                        "`runtime_env`."
```
Comment (edoakes) [inline python/ray/serve/utils.py:378]:
```suggestion
                        "the `runtime_env`."
```

File: python/ray/serve/http_adapters.py [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+
+    arr = np.array(payload.array)
+    if payload.shape:
+        arr = arr.reshape(*payload.shape)
+    if payload.dtype:
+        arr = arr.astype(payload.dtype)
+    return arr
+
+
+@require_packages(["PIL", "numpy"])
```
Comment (edoakes) [inline python/ray/serve/http_adapters.py:52]: nice, this decorator is great!

File: python/ray/serve/http_adapters.py [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+def image_to_databatch(img: bytes = File(...)) -> DataBatchType:
+    """Accept a PIL readable file from HTTP Form, returns array."""
```
Comment (edoakes) [inline python/ray/serve/http_adapters.py:54]:
```suggestion
    """Accepts a PIL-readable file from an HTTP form and converts it to a data batch type."""
```

File: python/ray/serve/http_adapters.py (hunk 2 above)
Comment (simon-mo) [reply python/ray/serve/http_adapters.py:31]: tried. there are like 158 of them. User will get a nice validation error from numpy anyway and with 158 types on documentation won't be very useful.
```
>>> import numpy as np
>>> np.sctypeDict.keys()
dict_keys(['?', 0, 'byte', 'b', 1, 'ubyte', 'B', 2, 'short', 'h', 3, 'ushort', 'H', 4, 'i', 5, 'uint', 'I', 6, 'intp', 'p', 7, 'uintp', 'P', 8, 'long', 'l', 'L', 'longlong', 'q', 9, 'ulonglong', 'Q', 10, 'half', 'e', 23, 'f', 11, 'double', 'd', 12, 'longdouble', 'g', 13, 'cfloat', 'F', 14, 'cdouble', 'D', 15, 'clongdouble', 'G', 16, 'O', 17, 'S', 18, 'unicode', 'U', 19, 'void', 'V', 20, 'M', 21, 'm', 22, 'bool8', 'Bool', 'b1', 'int64', 'Int64', 'i8', 'uint64', 'Uint64', 'u8', 'float16', 'Float16', 'f2', 'float32', 'Float32', 'f4', 'float64', 'Float64', 'f8', 'float128', 'Float128', 'f16', 'complex64', 'Complex32', 'c8', 'complex128', 'Complex64', 'c16', 'complex256', 'Complex128', 'c32', 'object0', 'Object0', 'bytes0', 'Bytes0', 'str0', 'Str0', 'void0', 'Void0', 'datetime64', 'Datetime64', 'M8', 'timedelta64', 'Timedelta64', 'm8', 'UInt64', 'int32', 'Int32', 'i4', 'uint32', 'UInt32', 'u4', 'int16', 'Int16', 'i2', 'uint16', 'UInt16', 'u2', 'int8', 'Int8', 'i1', 'uint8', 'UInt8', 'u1', 'complex_', 'int0', 'uint0', 'single', 'csingle', 'singlecomplex', 'float_', 'intc', 'uintc', 'int_', 'longfloat', 'clongfloat', 'longcomplex', 'bool_', 'bytes_', 'string_', 'str_', 'unicode_', 'object_', 'int', 'float', 'complex', 'bool', 'object', 'str', 'bytes', 'a'])
```

File: python/ray/serve/http_adapters.py (hunk 1 above)
Comment (simon-mo) [reply python/ray/serve/http_adapters.py:24]: it's just nested list, I will reword

File: python/ray/serve/http_adapters.py (hunk 3 above)
Comment (simon-mo) [reply python/ray/serve/http_adapters.py:18]: It's not specific to numpy. the input type is just nested list

File: python/ray/serve/http_adapters.py (hunk 2 above)
Comment (edoakes) [reply python/ray/serve/http_adapters.py:34]: I see. Can you update the description to specify that this is using the numpy dypte?

File: python/ray/serve/http_adapters.py [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+    if payload.shape:
+        arr = arr.reshape(*payload.shape)
+    if payload.dtype:
+        arr = arr.astype(payload.dtype)
+    return arr
+
+
+@require_packages(["PIL"])
+def image_to_databatch(img: bytes = File(...)) -> DataBatchType:
+    """Accepts a PIL-readable file from an HTTP form and converts it to a data batch type."""
```
Comment (edoakes) [inline python/ray/serve/http_adapters.py:54]:
```suggestion
    """Accepts a PIL-readable file from an HTTP form and converts it to a DataBatchType."""
```

Sorry, this is my own doing lol, didn't match above

**Review by edoakes — APPROVED** [review]

**Pull request merged**